sdks/python/apache_beam/transforms/userstate.py

```diff
@@ -184,29 +186,6 @@ def to_runner_api(self, context, key_coder, window_coder):
             coders._TimerCoder(key_coder, window_coder)))
 
 
-# TODO(BEAM-9602): Provide support for dynamic timer.
-class TimerFamilySpec(object):
```

Comment:
If one DoFn defines one `TimerSpec` and uses it in both `TimerParam` and `TimerFamilyParam`, we will have collisions on this spec, right?  And in this case, there is no way to register one callback for timer `TimerSpec and one callback for timer family `TimerSpec` since we are using maps in in implemetation.

I'm leaning to keep `TimerFamilySpec` and use `TimerParam` for both `TimerSpec` and `TimerFamilySpec`, where the `prefix` field will help us reduce the collision.

Reply:
Theoretically it should not supported to use same TimerSpec for different DoFnParam so I'm less worried about collision, if it is not apparent to user we can also add validation checks.

Reply:
I'm curious why do you choose `TimerSpec` -> `TimerParam`/`TimerFamilyParam` approach over `TimerSpec`/`TimerFamilySpec` -> `TimerParam`? Overall the API user is using `TimerSpec` to register `on_timer` callback, right?

Reply:
My main consideration is that the DoFnParam here implies the API, e.g. if using `a_timer=DoFn.TimerParam(timer_spec)` it translate to the old timer, if using `a_timer=DoFn.TimerFamilyParam(timer_spec`) it translates to a dynamic timer with which user can call `.get('tag').set()`, as for the timer spec it's main purpose is just to specify the name and time domain, eventually they are all mapped to https://github.com/apache/beam/blob/master/model/pipeline/src/main/proto/beam_runner_api.proto#L542, since we only have one model for the timer spec in beam runner api, I think there's also some advantages to just using one kind of timer spec in sdk.

Reply:
Thanks for the explanation! It sounds more like an API decision since implementation details are mostly the same.
My preference is:
```python
    class TimerDoFn(beam.DoFn):
      timer_spec = userstate.TimerSpec('timer', userstate.TimeDomain.WATERMARK)
      timer_family_spec = userstate.TimerFamilySpec('timer', userstate.TimeDomain.WATERMARK)
      def process(self,
                  element,
                  timer=beam.DoFn.TimerParam(timer_spec),
                  timer_map=beam.DoFn.TimerParam(timer_family_spec)):

      @userstate.on_timer(timer_spec)
      def process_timer(
          self,
          ts=beam.DoFn.TimestampParam,
          timer=beam.DoFn.TimerParam(timer_spec)):

      @userstate.on_timer(timer_family_spec)
      def process_timer(
          self,
          ts=beam.DoFn.TimestampParam,
          timer=beam.DoFn.TimerParam(timer_family_spec)):
``` 
And you are proposing the API like:
```python
    class TimerDoFn(beam.DoFn):
      timer_spec = userstate.TimerSpec('timer', userstate.TimeDomain.WATERMARK)
      timer_family_spec = userstate.TimerSpec('timer', userstate.TimeDomain.WATERMARK)
      def process(self,
                  element,
                  timer=beam.DoFn.TimerParam(timer_spec),
                  timer_map=beam.DoFn.TimerFamilyParam(timer_family_spec)):

      @userstate.on_timer(timer_spec)
      def process_timer(
          self,
          ts=beam.DoFn.TimestampParam,
          timer=beam.DoFn.TimerParam(timer_spec)):

      @userstate.on_timer(timer_family_spec)
      def process_timer(
          self,
          ts=beam.DoFn.TimestampParam,
          timer=beam.DoFn.TimerFamilyParam(timer_family_spec)):
```
And within your preferred API, the suer is not allowed to use the same name for both specs unless you want to additional logic around timer param to add prefix or just throw validation errors.

Do you think it's a good idea to ask options on dev list?

Reply:
Sounds good, I'll start a thread on dev list.